server/src/main/java/org/apache/lucene/index/ShuflleForcedMergePolicy.java

```diff
@@ -0,0 +1,119 @@
+ * and {@link MergePolicy#findForcedDeletesMerges}. This allows time-based indices, that usually have the eldest documents
+ * first, to be efficient at finding the most recent documents too.
+ */
+public class ShuflleForcedMergePolicy extends FilterMergePolicy {
+    private static final String SHUFFLE_MERGE_KEY = "es.shuffle_merge";
+
+    public ShuflleForcedMergePolicy(MergePolicy in) {
+        super(in);
+    }
+
+    /**
+     * Return <code>true</code> if the provided reader was merged with interleaved segments.
+     */
+    public static boolean isInterleavedSegment(LeafReader reader) {
+        SegmentReader segReader = Lucene.segmentReader(reader);
+        return segReader.getSegmentInfo().info.getDiagnostics().containsKey(SHUFFLE_MERGE_KEY);
+    }
+
+
+    @Override
+    public MergeSpecification findForcedDeletesMerges(SegmentInfos segmentInfos, MergeContext mergeContext) throws IOException {
+        return wrap(in.findForcedDeletesMerges(segmentInfos, mergeContext));
+    }
+
+    @Override
+    public MergeSpecification findForcedMerges(SegmentInfos segmentInfos, int maxSegmentCount,
+                                               Map<SegmentCommitInfo, Boolean> segmentsToMerge,
+                                               MergeContext mergeContext) throws IOException {
+        return wrap(in.findForcedMerges(segmentInfos, maxSegmentCount, segmentsToMerge, mergeContext));
+    }
+
+    private MergeSpecification wrap(MergeSpecification mergeSpec) throws IOException {
+        if (mergeSpec == null) {
+            return null;
+        }
+        MergeSpecification newMergeSpec = new MergeSpecification();
+        for (OneMerge toWrap : mergeSpec.merges) {
+            List<SegmentCommitInfo> newInfos = interleaveList(toWrap.segments);
+            newMergeSpec.add(new OneMerge(newInfos) {
+                @Override
+                public CodecReader wrapForMerge(CodecReader reader) throws IOException {
+                    return toWrap.wrapForMerge(reader);
+                }
+
+                @Override
+                public void setMergeInfo(SegmentCommitInfo info) {
+                    // Record that this merged segment is current as of this schemaGen:
+                    Map<String, String> copy = new HashMap<>(info.info.getDiagnostics());
+                    copy.put(SHUFFLE_MERGE_KEY, "");
+                    info.info.setDiagnostics(copy);
+                    super.setMergeInfo(info);
+                }
+            });
+        }
+
+        return newMergeSpec;
+    }
+
+    // Return a new list that sort segments of the original one by name (older first)
+    // and then interleave them to colocate oldest and most recent segments together.
+    private List<SegmentCommitInfo> interleaveList(List<SegmentCommitInfo> infos) throws IOException {
+        List<SegmentCommitInfo> newInfos = new ArrayList<>(infos.size());
+        Collections.sort(infos, Comparator.comparing(a -> a.info.name));
```

Comment:
I think we should avoid changing `infos` in place.

Reply:
Making a copy would also help ensure that the list supports random-access.

Reply:
++, I pushed https://github.com/elastic/elasticsearch/pull/48533/commits/aae5c3013388034f0fb9e3d6d9beaa50c0f6e09a